source/containerimage/ocilayout.go

```diff
@@ -0,0 +1,195 @@
+package containerimage
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/reference"
+	"github.com/containerd/containerd/remotes"
+	"github.com/moby/buildkit/session"
+	sessioncontent "github.com/moby/buildkit/session/content"
+	"github.com/moby/buildkit/source"
+	digest "github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// GetOCILayoutResolver gets a resolver to an OCI layout for a specified scope from the pool
+func GetOCILayoutResolver(layoutPath string, sm *session.Manager, sessionID string, g session.Group) *OCILayoutResolver {
+	r := &OCILayoutResolver{
+		path:      layoutPath,
+		sm:        sm,
+		sessionID: sessionID,
+		g:         g,
+	}
+	return r
+}
+
+type OCILayoutResolver struct {
+	remotes.Resolver
+	path      string
+	sm        *session.Manager
+	sessionID string
+	g         session.Group
+	is        images.Store
+	mode      source.ResolveMode
+}
+
+// WithImageStore returns new resolver that can also resolve from local images store
+func (r *OCILayoutResolver) WithImageStore(is images.Store, mode source.ResolveMode) *OCILayoutResolver {
+	r2 := *r
+	r2.is = is
+	r2.mode = mode
+	return &r2
+}
+
+// Fetcher returns a new fetcher for the provided reference.
+func (r *OCILayoutResolver) Fetcher(ctx context.Context, ref string) (remotes.Fetcher, error) {
+	return r, nil
+}
+
+// Fetch get an io.ReadCloser for the specific content
+func (r *OCILayoutResolver) Fetch(ctx context.Context, desc ocispecs.Descriptor) (io.ReadCloser, error) {
+	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	sessionID := r.sessionID
+
+	caller, err := r.sm.Get(timeoutCtx, sessionID, false)
+	if err != nil {
+		return r.fetchWithAnySession(ctx, desc)
```

Comment:
Not sure if we need this. I think we should always require a fixed session.

Reply:
Doesn't matter to me. I just copied the format we had for local. I don't mind getting rid of it and just returning an error if it cannot get the session. Your call.